test/test_iterdatapipe.py

```diff
@@ -993,6 +993,159 @@ def test_drop_iterdatapipe(self):
+        slice_dp = input_dp.slice(["a", "b"])
+        self.assertEqual([{"a": 1, "b": 2}, {"a": 3, "b": 4}, {"a": 5, "b": 6}], list(slice_dp))
+
+        # __len__ Test:
+        input_dp = IterableWrapper([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
+        slice_dp = input_dp.slice(0, 2)
+        self.assertEqual(3, len(slice_dp))
+
+        # Reset Test:
+        n_elements_before_reset = 2
+        input_dp = IterableWrapper([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
+        slice_dp = input_dp.slice([2])
+        expected_res = [[2], [5], [8]]
+        res_before_reset, res_after_reset = reset_after_n_next_calls(slice_dp, n_elements_before_reset)
+        self.assertEqual(expected_res[:n_elements_before_reset], res_before_reset)
+        self.assertEqual(expected_res, res_after_reset)
+
+    def test_flatten_iterdatapipe(self):
+        # tuple tests
+
+        # Functional Test: flatten for an index
+        input_dp = IterableWrapper([(0, 1, (2, 3)), (4, 5, (6, 7)), (8, 9, (10, 11))])
+        flatten_dp = input_dp.flatten(2)
+        self.assertEqual([(0, 1, 2, 3), (4, 5, 6, 7), (8, 9, 10, 11)], list(flatten_dp))
+
+        # Functional Test: flatten for list of indices
+        input_dp = IterableWrapper([((0, 10), 1, (2, 3)), ((4, 14), 5, (6, 7)), ((8, 18), 9, (10, 11))])
+        flatten_dp = input_dp.flatten([0, 2])
+        self.assertEqual([(0, 10, 1, 2, 3), (4, 14, 5, 6, 7), (8, 18, 9, 10, 11)], list(flatten_dp))
+
+        # Functional Test: flatten all iters in the datapipe one level (no argument)
+        input_dp = IterableWrapper([(0, (1, 2)), (3, (4, 5)), (6, (7, 8))])
+        flatten_dp = input_dp.flatten()
+        self.assertEqual([(0, 1, 2), (3, 4, 5), (6, 7, 8)], list(flatten_dp))
+
+        # list tests
+
+        # Functional Test: flatten for an index
+        input_dp = IterableWrapper([[0, 1, [2, 3]], [4, 5, [6, 7]], [8, 9, [10, 11]]])
+        flatten_dp = input_dp.flatten(2)
+        self.assertEqual([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]], list(flatten_dp))
+
+        # Functional Test: flatten for list of indices
+        input_dp = IterableWrapper([[[0, 10], 1, [2, 3]], [[4, 14], 5, [6, 7]], [[8, 18], 9, [10, 11]]])
+        flatten_dp = input_dp.flatten([0, 2])
+        self.assertEqual([[0, 10, 1, 2, 3], [4, 14, 5, 6, 7], [8, 18, 9, 10, 11]], list(flatten_dp))
+
+        # Functional Test: flatten all iters in the datapipe one level (no argument)
+        input_dp = IterableWrapper([[0, [1, 2]], [3, [4, 5]], [6, [7, 8]]])
+        flatten_dp = input_dp.flatten()
+        self.assertEqual([[0, 1, 2], [3, 4, 5], [6, 7, 8]], list(flatten_dp))
+
+        # Functional Test: string test, flatten all iters in the datapipe one level (no argument)
+        input_dp = IterableWrapper([["zero", ["one", "2"]], ["3", ["4", "5"]], ["6", ["7", "8"]]])
+        flatten_dp = input_dp.flatten()
+        self.assertEqual([["zero", "one", "2"], ["3", "4", "5"], ["6", "7", "8"]], list(flatten_dp))
+
```

Comment:
Nit: Probably include test and mention in the doc, that we are not supporting nesting levels > 1